### PR TITLE
LightningCLI support for argument links applied on instantiation

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

From discussions mostly in slack it became clear that there are common cases in which parameters that are only available after instantiating the data module class are needed to instantiate the model. To achieve the user had to subclass LightningCLI in a non-trivial way. With this PR these cases can be implemented in a much simpler way. The concept of argument linking is extended so that they can be applied on instantiation. It use is explained in the change of lightning_cli.rst.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary)
- [X] Did you write any new necessary tests? (not for typos and docs)
- [X] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/common/lightning_cli.rst [hunk 1]
```diff
@@ -402,6 +406,23 @@ The linking of arguments is observed in the help of the tool, which for this exa
         model.batch_size <-- data.batch_size
                               Number of samples in a batch (type: int)
 
+Sometimes a parameter value is only available after class instantiation. An example could be a data module for a
+classification task which counts the number of classes. And the number of classes is required to instantiate the model.
```
Comment (carmocca) [inline docs/source/common/lightning_cli.rst:410]:
```suggestion
Sometimes a parameter value is only available after class instantiation. An example could be that your model requires the number of classes to instantiate its fully connected layer (for a classification task) but the value is not available until the data module has been instantiated.
```

File: docs/source/common/lightning_cli.rst [hunk 2]
```diff
@@ -402,6 +406,23 @@ The linking of arguments is observed in the help of the tool, which for this exa
...
+The code below illustrates this.
```
Comment (carmocca) [inline docs/source/common/lightning_cli.rst:411]:
```suggestion
The code below illustrates how to address this.
```

File: pytorch_lightning/utilities/cli.py [hunk 3]
```diff
@@ -212,27 +217,11 @@ def before_instantiate_classes(self) -> None:
...
-        else:
-            self.datamodule = self.datamodule_class(**self.config_init.get('data', {}))
-
-    def instantiate_model(self) -> None:
-        """Instantiates the model using self.config_init['model']"""
-        if self.subclass_mode_model:
-            self.model = self.config_init['model']
-        else:
-            self.model = self.model_class(**self.config_init.get('model', {}))
-
```
Comment (carmocca) [inline pytorch_lightning/utilities/cli.py:235]: Why are these gone?

We will need them in Flash to instantiate the datamodule from a `classmethod`

**Review by carmocca — COMMENTED** [review]
Still draft? What's missing?

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (mauvilsa) [reply pytorch_lightning/utilities/cli.py:235]: It is necessary to remove them to make it possible to have links that are applied at instantiation time. Based on the links the order of instantiation is determined. There is a loop, the first class is instantiated followed by applying the links that make use of that class, go to the second class and repeat.

Regarding flash I would need to know more about what is the situation there. Anyway note that currently these methods are inconsistent. In subclass mode the class has already been instantiated, whereas in normal mode the init args are used to instantiate. Don't know how this is used in flash but it probably has this subclass thing broken. Depending on the exact need in flash we can see how to do it. But it is till best to remove these methods.

**Review by awaelchli — APPROVED** [review]
smart!

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:235]: In Flash, `DataModule`s can be created using different `classmethods`.

Some examples are `from_folders`, `from_csv`, `from_tensors`, ...

https://github.com/PyTorchLightning/lightning-flash/blob/a047330ba75486355378f22cbebfd053c3d63c08/flash_examples/finetuning/image_classification.py#L28

A `DataModule` can support multiple of these.

But actually, I was being dumb, we just need to support the following:


```python
import torch
from torch.utils.data import Dataset, DataLoader

import pytorch_lightning as pl
from pytorch_lightning.utilities.cli import LightningCLI


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class MyDataModule(pl.LightningDataModule):
    @classmethod
    def from_thing(cls):
        return cls()

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 20))


class MyModel(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


cli = LightningCLI(MyModel, MyDataModule.from_thing, trainer_defaults={'fast_dev_run': True})
```

So passing a function to create the datamodule.

So there's no need to subclass and override `instantiate_datamodule`.

Should be easy right? Just need to call `add_function_arguments` if the passed input is a function instead of a class

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (mauvilsa) [reply pytorch_lightning/utilities/cli.py:235]: Yes, could be. However, with `add_function_arguments` the instantiation links wouldn't work. Though I do think something simple could be done to make them work.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:235]: Do you want to open a PR with the improvement?
